### ISSUE TITLE
Support Global Filters

### ISSUE THREAD

**Issue by jcloutz-radix** [prisma/prisma2#1187]
It would be great if Photon had the ability to apply global filters to models. It has many uses, but primarily for me it would be nice to be able to automatically apply a tenant id filter to queries without having to do it explicitly everywhere the model is used.

This is a massively helpful feature that I haven't seen in any node.js ORM's, and coming from .Net and Entity Framework I truly miss the flexibility it adds.

Additionally it would be amazing if the filters could be applied conditionally so that they can be turned disabled temporarily if necessary. 

This would of course depend on a clean way to isolate and handle the request context.

**Comment by matthewmueller** [comment]
@jcloutz-radix I'm not familiar with global filters. Does this mean that you create a model with a given ID already applied?

> .Net and Entity Framework I truly miss the flexibility it adds. 

Do you have some documentation or examples that I could take a look at?

**Comment by jcloutz-radix** [comment]
@matthewmueller  #2851 pretty much sums what I would like to see. But it would be nice if it were to go beyond a where clause and have the ability to automatically perform joins on other tables. There are a lot of instances where this would be helpful.

- Automatically filtering by a tenant identifier within the table
- Performing a join on a relation to determine the tenant 
- Joining on permission or ownership tables to to determine if a user is even allowed to access a specific resource.

Here are the docs for entity framework and how it accomplishes it, https://docs.microsoft.com/en-us/ef/core/querying/filters. This takes advantage of .net's integrated dependency injection to pass the authenticated user info into the db context for every request.

However, similar results could probably achieved by allowing prisma client to receive a per-request user-defined context object. The context could contain any relevant data required, user_id, tenant_id, scopes, etc... That object would then be used to provide values to the filter functions that modify the query.

**Comment by pantharshit00** [comment]
@jcloutz-radix I am not sure if you are already familiar but there is a community project right now that can give you multi tenant features: https://github.com/errorname/prisma-multi-tenant

**Comment by jcloutz-radix** [comment]
@pantharshit00 yes, i have played with that in the past, but it's aimed at a multi-database setup. Global filters would add a lot of options that go beyond just filtering on tenant.

**Comment by Val-istar-Guo** [comment]
it is very convenient to implement soft delete function. TypeORM have the same issue:

https://github.com/typeorm/typeorm/issues/1601

**Comment by Papooch** [comment]
You can take inspiration from the concept of *Scopes* from *Sequelize*: https://sequelize.org/master/manual/scopes.html
I absolutely LOVE that feature, however, I'm not sure if it's possible to strongly type something like that.

**Comment by Hatko** [comment]
It's suggested in the Prisma documentation to use middleware for the purpose of [soft deleting](https://www.prisma.io/docs/concepts/components/prisma-client/middleware/soft-delete-middleware#option-2-use-middleware-to-determine-the-behavior-of-readupdate-queries-for-deleted-records). Probably can be used for global filtering as well

**Comment by igalklebanov** [comment]
Hey 👋 

Closing this as its dusty, niche and can be achieved with extensions.
